File: solutions/203. Remove Linked List Elements.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
​
class Solution:
    def removeElements(self, head: Optional[ListNode], val: int) -> Optional[ListNode]:
        
        return self.sol2(head, val)
        
    # O(N), O(1)  
    def sol1(self, head: Optional[ListNode], val: int) -> Optional[ListNode]:
        
        pre = ListNode()
        pre.next = head
        curr = pre
        
        while curr.next:
            if curr.next.val == val: curr.next = curr.next.next
            else: curr = curr.next
        
        return pre.next
    
    # O(N), O(N)(stack space) 
    def sol2(self, head: Optional[ListNode], val: int) -> Optional[ListNode]:
        
        if not head: return None
        
        if head.val == val:
            head = self.sol2(head.next, val)
        else:
            head.next = self.sol2(head.next, val)
        
        return head
        
​

File: solutions/203. Remove Linked List Elements.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
​
public class Solution {
    public ListNode RemoveElements(ListNode head, int val)
    {
        ListNode pre = new ListNode(0, head);
        var curr = pre;
​
        while (curr.next != null)
        {
            if (curr.next.val == val)
            {
                curr.next = curr.next.next;
            }
            else
            {
                curr = curr.next;
            }
        }
        return pre.next;
    }
}
